thunder/dynamo/utils.py

```diff
@@ -512,3 +512,45 @@ def checkpoint_converter(gm: torch.fx.GraphModule, sub_gm: torch.fx.GraphModule)
                 else:
                     function_module = getattr(gm, n.args[0].name)
                 _checkpoint_function_converter(function_module)
+
+
+def remove_empty_autocast(graph_module: torch.fx.GraphModule) -> torch.fx.GraphModule:
+    """
+    Function to remove empty autocast regions from GraphModule.
+
+    Dynamo can provide empty autocast regions in which case, it is more performant to remove them
+    from the graph than to compile them and pay the cost of calling a wrapped optimized function
+    which does nothing.
+
+    Args:
+        graph_module: Graph module to which this pass is applied.
+
+    """
+
+    empty_autocast_removed_graph_module = copy.deepcopy(graph_module)
+
+    # Dummy init node.
+    prev_node = torch.fx.node.Node(graph_module.graph, "start_node", "call_function", lambda: None, None, None)
+    nodes_to_erase = []
+    for node in empty_autocast_removed_graph_module.graph.nodes:
+        # Due to these functions mapping regions created by context manager,
+        # previous `_enter_autocast` will always correspond with current `_exit_autocast`.
+        if (
+            prev_node.target == torch.amp.autocast_mode._enter_autocast
+            and node.target == torch.amp.autocast_mode._exit_autocast
```

Comment:
What other enter and exits may appear in the Dynamo graph?

Reply:
Besides {_enter/_exit}_autocast, from [trace_rules.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/trace_rules.py), I could find
{_enter/_exit}_dual_level for forward mode auto-diff, {_enter/_exit}_inference_mode for inference mode.

Besides these, looking at [ctx_manager.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/ctx_manager.py), there could also be  `Stream`, `Grad`, `Device` related context managers which are mapped to something like the following

```python

old_stream = torch.cuda.get_current_stream()
new_stream = torch.cuda.Stream()
set_stream(new_stream)
# Some operations
set_stream(old_stream)

```


Refs:
https://github.com/pytorch/pytorch/blob/781c68c86549aa2fd155f92afb0406b7577ab31c/torch/_dynamo/trace_rules.py#L550-L552
https://github.com/pytorch/pytorch/blob/781c68c86549aa2fd155f92afb0406b7577ab31c/torch/_dynamo/trace_rules.py#L2393-L2394

Reply:
Thank you for the links! I created an issue from this comment: https://github.com/Lightning-AI/lightning-thunder/issues/1412.